### DISCUSSION TITLE
Event prover

### DISCUSSION THREAD

**Discussion by cenwadike** [Near-One/rainbow-token-connector#103] [Q&A] [answered]
Can someone please explain in simple words what this code does. link below

https://github.com/aurora-is-near/rainbow-token-connector/blob/master/bridge-token-factory/src/prover.rs#:~:text=%7D-,impl%20EthEvent%20%7B,%7D,-fn%20long_signature(name

I understand `pub fn from_log_entry_data` populate the fields of Proof which is digested by a cross contract call to `ext_prover` contract.
I don't understand how `input` field in `Event` struct is populated.
I also don't really understand `rlp` role.
for the sake of completeness of knowledge, can the link of  `ext_prover` be provided.

**Comment by mfornet** [comment, marked as answer]
> for the sake of completeness of knowledge, can the link of ext_prover be provided.

Here is the code for Prover: [ext_prover::verify_log_entry](https://github.com/aurora-is-near/rainbow-bridge/blob/master/contracts/near/eth-prover/src/lib.rs#L108). It is used this way for cross contract calls on NEAR, in this case it calls the prover on Mainnet.

> I also don't really understand rlp role.

[RLP](https://eth.wiki/fundamentals/rlp) is the standard for binary serialization in Ethereum, and we keep it as is for simplicity at review time! It is relevant for data that should be hashed, and it is expected that the content to match byte by byte to the original.

> I don't understand how input field in Event struct is populated.

It is the data related to Ethereum logs (which in this context are called Events). Check details on [the implementation](https://github.com/paritytech/ethabi/blob/master/ethabi/src/event_param.rs#L19-L27). It basically have the definition of the event (Name, Signatures, Indexed Flag).

**Reply by cenwadike** [reply]
Thanks for the prompt response
